user_test.go

```diff
@@ -0,0 +1,203 @@
+package gomatrixserverlib_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/matrix-org/gomatrixserverlib"
+)
+
+const defaultDomain = "domain"
+const defaultLocalpart = "localpart"
+
+func TestEmptyFails(t *testing.T) {
+	_, err := gomatrixserverlib.NewUserID("", false)
+	if err == nil {
+		t.Fatalf("empty userID is not valid, it shouldn't parse")
+	}
+}
+
+func TestBasicValidSucceeds(t *testing.T) {
+	localpart := defaultLocalpart
+	domain := defaultDomain
+	raw := fmt.Sprintf("@%s:%s", localpart, domain)
+
+	userID, err := gomatrixserverlib.NewUserID(raw, false)
+
+	if err != nil {
+		t.Fatalf("valid userID should not fail")
+	}
+	if userID.Local() != localpart {
+		t.Fatalf("Localpart - Expected: %s Actual: %s ", localpart, userID.Local())
+	}
+	if userID.Domain() != gomatrixserverlib.ServerName(domain) {
+		t.Fatalf("Domain - Expected: %s Actual: %s ", gomatrixserverlib.ServerName(domain), userID.Domain())
+	}
+	if userID.Raw() != raw {
+		t.Fatalf("Raw - Expected: %s Actual: %s ", raw, userID.Raw())
+	}
+}
+
+func TestExtensiveLocalpartSucceeds(t *testing.T) {
+	localpart := "abcdefghijklmnopqrstuvwxyz0123456789._=-/"
+	domain := defaultDomain
+	raw := fmt.Sprintf("@%s:%s", localpart, domain)
+
+	userID, err := gomatrixserverlib.NewUserID(raw, false)
+
+	if err != nil {
+		t.Fatalf("valid userID should not fail")
+	}
+	if userID.Local() != localpart {
+		t.Fatalf("Localpart - Expected: %s Actual: %s ", localpart, userID.Local())
+	}
+	if userID.Domain() != gomatrixserverlib.ServerName(domain) {
+		t.Fatalf("Domain - Expected: %s Actual: %s ", domain, userID.Domain())
+	}
+	if userID.Raw() != raw {
+		t.Fatalf("Raw - Expected: %s Actual: %s ", raw, userID.Raw())
+	}
+}
+
+func TestExtensiveLocalpartHistoricalSucceeds(t *testing.T) {
+	localpart := "!\"#$%&'()*+,-./0123456789;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~"
+	domain := defaultDomain
+	raw := fmt.Sprintf("@%s:%s", localpart, domain)
+
+	userID, err := gomatrixserverlib.NewUserID(raw, true)
+
+	if err != nil {
+		t.Fatalf("valid userID should not fail")
+	}
+	if userID.Local() != localpart {
+		t.Fatalf("Localpart - Expected: %s Actual: %s ", localpart, userID.Local())
+	}
+	if userID.Domain() != gomatrixserverlib.ServerName(domain) {
+		t.Fatalf("Domain - Expected: %s Actual: %s ", domain, userID.Domain())
+	}
+	if userID.Raw() != raw {
+		t.Fatalf("Raw - Expected: %s Actual: %s ", raw, userID.Raw())
+	}
+}
+
+func TestDomainWithPortSucceeds(t *testing.T) {
```

Comment:
Table-driven tests would be preferable here, for all these tests. 

https://dave.cheney.net/2019/05/07/prefer-table-driven-tests